README.md

```diff
@@ -4,32 +4,123 @@
+* [Demos](#demos)
+* [Developers](#developers)
+* [Supported Kubernetes versions](#supported-kubernetes-versions)
+* [Related code](#related-code)
 
 ## About
 
-This repository contains a set of Kubernetes plugins that enable you to use
-Intel devices.
+This repository contains a framework for developing plugins for the Kubernetes
+[device plugins framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/),
+along with a number of device plugin implementations utilising that framework.
 
-Prerequisites for building and running Intel device plugins include:
+## Prerequisites
 
-- Intel hardware
-- Fully configured [Kubernetes cluster]
-- Working [Go environment]
+Prerequisites for building and running these device plugins include:
+
+- Appropriate hardware
+- A fully configured [Kubernetes cluster]
+- A working [Go environment]
+
+## Plugins
+
+The below sections detail existing plugins developed using the framework.
+
+### GPU device plugin
+
+The [GPU device plugin](cmd/gpu_plugin/README.md) supports Intel
+[GVT-d](https://github.com/intel/gvt-linux/wiki/GVTd_Setup_Guide) device passthrough
+and acceleration using GPUs of the following hardware families:
+
+- Integrated GPUs within Intel Core processors
+- Intel Xeon processors
+- Intel Visual Compute Accelerator (Intel VCA)
+
+The demo subdirectory contains both a [GPU plugin demo video](demo#intel-gpu-device-plugin-demo-video)
+as well as code for an OpenCL [FFT demo](demo/ubuntu-demo-opencl).
+
+### FPGA device plugin
+
+The [FPGA device plugin](cmd/fpga_plugin/README.md) supports Intel Arria 10 and Stratix 10
+FPGA passthrough. It supports the [OPAE](https://01.org/opae) mode.
+
+The FPGA plugin comes as three parts. Start with the [device plugin](#device-plugin), and then
+examine the [admissing controller](#admission-controller) and the
+[CRIO-O prestart hook](#cri-o-prestart-hook).
+
+> **FIXME** - unclear where the [fpga_tool](cmd/fpga_tool) comes into play. We should document this.
+
+The demo subdirectory contains a [video](demo#intel-fpga-device-plugin-demo-video) showing deployment
+and use of the FPGA plugin. Sources relating to the demo can be found in the
+[opae-nlb-demo](demo/opae-nlb-demo) subdirectory.
+
+#### device plugin
+
+The [FPGA device plugin](cmd/fpga_plugin/README.md) is responsible for discovering and reporting FPGA
+devices to `kubelet`. During the allocation phase, it instructs `kubelet` about device nodes to be
+accessed by the container and sets the annotation that is later on passed to CRI to trigger programming
+by the CRI-O hook.
+
+#### admission controller
+
+The [FPGA admission controller webhook](cmd/fpga_admissionwebhook/README.md) is responsible for performing
+mapping from user-friendly function IDs to the Interface ID and Bitstream ID that are required for FPGA
+programming by the CRI-O hook. Mappings are stored in namespaced custom resource definition (CRD) objects,
```

Comment:
We should document the two modes we support and the fact that the other mode has more dependencies: the orchestration programmed mode depens on CRI-O and the prestart hooks.

Reply:
so the two modes are 'preprogrammed mode' and 'orchestrated mode'? Any more details or the wording for what to insert here?

I'll note, I have considered if these fairly verbose descriptions would be better placed in the submodule (fpga plugin) docs themselves. But, for now, we can leave them here.

Reply:
Yes, two modes are 'preprogrammed' and 'orchestrated' or 'orchestration programmed'. Preprogrammed mode allows to use functions already programmed into the FPGA device. It's simple and restrictive and it doesn't depend on CRI-O i.e. Docker can be used as a runtime. When there is no need to reprogram devices this mode is definitely recommended to use.

Orchestrated mode is more complex as it allows dynamic reprogramming of the devices. It uses CRI-O as this is the only runtime that allows running prestart hooks before starting containers with a user workflow. Programming happens in the prestart hook installed by initcontainer.

Please, note that user workflow can't program devices in any of 2 modes mentioned above. This is done for security reasons. Programming of the devices in orchestration mode is done by orchestration, before the user workload starts.

Reply:
I've made the additions that there are two modes, and only orchestration mode requires the prestart hook (I'll push those changes here soon).
I have a changeset pending that updates all the docs in the cmd/fpga_* subfolders - so will restrict the details added here, and we can add them to the appropriate subfolder (rather than cluttering the top level).

Reply:
@grahamwhaley 

> It supports the [OPAE](https://01.org/opae) mode.

OPAE is a framework and only two pieces of software are used from it:
- [out of tree Linux device driver](https://opae.github.io/latest/docs/drv_arch/drv_arch.html)
- nlb0 and nlb3 utilities in the [demo image](https://github.com/intel/intel-device-plugins-for-kubernetes/tree/master/demo/opae-nlb-demo) for demonstration purposes only.

It would be also good to mention that [in-tree device driver](https://www.kernel.org/doc/html/latest/fpga/dfl.html) is also supported by the plugin.

Reply:
FPGA plugin also operates in 2 modes: preprogrammed and orchestration programmed. In can work with Docker or any other runtime (CRI-O, Containerd) in preprogrammed mode, but requires CRI-O and prestart hook in orchestration programmed mode.

Reply:
@bart0sh @mythi - OK, what I'm going to do is trim these top level readme sections down to a minimum. I have a WIP tree that expands the FPGA specific READMEs in progress, and will PR that soon - and then we can add all the details about the FPGA to the FPGA specific docs.